protonmail/attachments.go

```diff
@@ -75,20 +75,33 @@ func (att *Attachment) GenerateKey(to []*openpgp.Entity) (*packet.EncryptedKey,
 // GenerateKey.
 //
 // signed is ignored for now.
-func (att *Attachment) Encrypt(ciphertext io.Writer, signed *openpgp.Entity) (cleartext io.WriteCloser, err error) {
+func (att *Attachment) Encrypt(ciphertext io.Writer, signature io.WriteCloser, signed *openpgp.Entity) (cleartext io.WriteCloser, err error) {
```

Comment:
Hmm. This API is a little scary, because we won't be done writing the signature to the supplied `io.WriteCloser` by the time this function returns…

Reply:
Maybe a better alternative would be to add a `Signature` method to `signatureWriter` that allows retrieving the generated signature. Would involve changing the `io.WriteCloser` return value.

Reply:
I've finally got some time to work on this. It's been a while, so I'm not sure about order of calls and it may be a bit overcomplicated or otherwise questionable.

That's a valid point. I've rewritten it in this way, but it required some trickery since `self` is passed in `io.WriteCloser` methods by value and we need to share `signatureWriter` instance there. Alternatively we may ditch `io.WriterCloser` compatibility altogether and pass the while structure as a context, but I don't know if it would be better

Reply:
sounds good to me, what do you think @emersion ?